File: 20250319/leetcode2610. 转换二维数组/main.js
// 给你一个整数数组 nums 。请你创建一个满足以下条件的二维数组：

// 二维数组应该 只 包含数组 nums 中的元素。
// 二维数组中的每一行都包含 不同 的整数。
// 二维数组的行数应尽可能 少 。
// 返回结果数组。如果存在多种答案，则返回其中任何一种。

// 请注意，二维数组的每一行上可以存在不同数量的元素。

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var findMatrix = function (nums) {
  const n = nums.length;
  const hash = new Map();
  let max = 1;
  for (const num of nums) {
    hash.set(num, (hash.get(num) || 0) + 1);
    max = Math.max(max, hash.get(num));
  }
  const ans = new Array(max).fill(0).map((v) => new Array(0).fill(0));
  for (let [key, value] of hash) {
    for (let i = 0; i < value; i++) ans[i].push(key);
  }
  return ans;
};

File: 20250319/leetcode2610. 转换二维数组/main.cpp
#include <cmath>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
  vector<vector<int>> findMatrix(vector<int> &nums) {
    unordered_map<int, int> hash;
    for (int x : nums) {
      hash[x]++;
    }
    vector<vector<int>> ans;
    while (!hash.empty()) {
      vector<int> arr;
      for (auto it = hash.begin(); it != hash.end();) {
        arr.push_back(it->first);
        it->second -= 1;
        if (it->second == 0) {
          it = hash.erase(it);
        } else {
          it++;
        }
      }
      ans.push_back(arr);
    }
    return ans;
  }
};

int main(void) {
  Solution solution;
  vector<int> nums = {1, 2, 3, 4, 4, 4, 5, 6, 1,  2,  1,  1,  2, 4,
                      1, 5, 6, 6, 6, 9, 8, 5, 11, 12, 12, 12, 11};
  auto ans = solution.findMatrix(nums);
  int n = ans.size();
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < ans[i].size(); j++) {
      cout << ans[i][j] << "\t";
    }
    cout << endl;
  }
}